src/forwarder.rs

```diff
@@ -0,0 +1,387 @@
+    where
+        P: Publisher,
+    {
+        while let Some(msg) = rx_state.recv().await {
+            let ipath = format!("/{}/status", msg.token);
+            let idata = msg.into();
+
+            if let Err(err) = publisher
+                .send(
+                    "io.edgehog.devicemanager.ForwarderSessionsState",
+                    &ipath,
+                    idata,
+                )
+                .await
+            {
+                error!("publisher send error, {err}");
+            }
+        }
+    }
+
+    /// Start a device forwarder instance.
+    pub fn handle_sessions(&mut self, astarte_event: AstarteDeviceDataEvent) {
+        let Some(idata) = Self::retrieve_astarte_data(astarte_event) else {
+            return;
+        };
+
+        // retrieve the Url that the device must use to open a WebSocket connection with a host
+        let cinfo = match retrieve_connection_info(idata) {
+            Ok(cinfo) => cinfo,
+            // error while retrieving the connection information from the Astarte data
+            Err(err) => {
+                error!("{err}");
+                return;
+            }
+        };
+
+        let bridge_url = match Url::try_from(&cinfo) {
+            Ok(url) => url,
+            Err(err) => {
+                error!("invalid url, {err}");
+                return;
+            }
+        };
+
+        // check if the remote terminal task is already running. if not, spawn a new task and add it to the collection
+        let tx_state = self.tx_state.clone();
+        let session_token = cinfo.session_token.clone();
+        self.get_running(cinfo).or_insert_with(||
+            // spawn a new task responsible for handling the remote terminal operations
+            Self::spawn_task(bridge_url, session_token, tx_state));
+    }
+
+    fn retrieve_astarte_data(
+        astarte_event: AstarteDeviceDataEvent,
+    ) -> Option<HashMap<String, AstarteType>> {
+        if astarte_event.path != "/request" {
+            error!("received data from an unknown path/interface: {astarte_event:?}");
+            return None;
+        }
+
+        let Aggregation::Object(idata) = astarte_event.data else {
+            error!("received wrong Aggregation data type");
+            return None;
+        };
+
+        Some(idata)
+    }
+
+    /// Remove terminated sessions and return the searched one.
+    fn get_running(&mut self, cinfo: SessionInfo) -> Entry<Key, JoinHandle<()>> {
+        // remove all finished tasks
+        self.tasks.retain(|_, jh| !jh.is_finished());
+
+        self.tasks.entry(Key(cinfo))
+    }
+
+    /// Spawn the task responsible for handling a device forwarder instance.
+    fn spawn_task(
+        bridge_url: Url,
+        session_token: String,
+        tx_state: Sender<SessionState>,
+    ) -> JoinHandle<()> {
+        tokio::spawn(Self::handle_session(bridge_url, session_token, tx_state))
+    }
```

Comment:
This can probabli be inlined

Reply:
by using `cargo fmt` it still remains the same